core/base/src/TApplication.cxx

```diff
@@ -156,12 +141,6 @@ TApplication::TApplication(const char *appClassName, Int_t *argc, char **argv,
    if (!gSystem)
       ::Fatal("TApplication::TApplication", "gSystem not initialized");
 
-   static Bool_t hasRegisterAtExit(kFALSE);
```

Comment:
humm .. if we don't use atexit, what is guaranteeing the call to EndOfProcessCleanups to be done early enough? (i.e. we are aiming **before** any library is unloaded).

Reply:
What do you mean by early enough? This PR makes it possible to run the atexit handlers at the destruction time of TCling. IIUC a bunch of atexit handlers were added exactly because we could not reply on TCling to clean up.

Reply:
Destruction time of TCling is *way* too late to run EndOfProcessCleanups, this *must* be run before any *user* library has been unloaded otherwise when the file try to store the attached object it may need to run code that has already been unloaded.  Ideally it would be 'right after' the return of the main function ... but since we can't get there all the time, the try to schedule the atexit as early as possible.

Reply:
When you say 'user library has been unloaded' you mean unloaded at shutdown time or `dlclosed` at any time?

Reply:
I mean " unloaded at shutdown time" as we assume that if the user chose to call (directly or indirectly) dlclose, it has 'first' dealt with all the objects or the related types ....

Reply:
and FYI, I don't think all use cases are covered in roottest and thus I would only removed this if there is a clear use case where it is getting in the way.

Reply:
Looking at the code I do not see how this call make things happen earlier as a few lines above it we had a use of gROOT which triggers (likely) unintentional initialization of ROOT (unless initialized in users code).

Would you mind to commit this as is and ask the experiments for feedback? In case of breakage I will add the relevant tests and revert that particular change?

Reply:
The experiment are the worse tester for this kind of problem .. because they are already do proper memory management and do not need to really on those helpers.

The challenge with this problem is that some point in the future some user will report a odd end of process problem that we may or may not easily recognize has due to those changes ....

Reply:
> use of gROOT which triggers (likely) 

Unless gROOT has already been used a long time ago (in the start of the process).   Eg.
```
#include "TApplication.h" 

static int rootversion = gROOT->GetVersionInt();

int main(int, char**)
{
   dlopen(somelibrary);
   auto app = new TApplication("app","title"); 

   return 0;
}
```
will behave differently (also depending of the content of the libraries hard-linked and dlopened).

And yes this example is contrived but it can easily happen implicitly (in particular because if the user does not create a TApplication, one will be created implicitly, for example triggered by a call to TROOT::ProcessLine).

Reply:
It is unclear at the moment to me how this may address this specific set of use cases. I will remove this commit. I'd be happy if we had concrete test cases describing the use case and the intent.

Reply:
The comment of `CallEndOfProcessCleanups` essentially says that `atexit` gets called after calling `ShutDown` at least on Ubuntu (thus the if-stmt to protect the code there).